core/routing_spray.go

```diff
@@ -0,0 +1,313 @@
+	} else {
+		metadata := sprayMetaData{
+			sent:            make([]bundle.EndpointID, 0),
+			remainingCopies: sprayl,
+		}
+		bs.bundleData[bp.ID()] = metadata
+		log.WithFields(log.Fields{
+			"bundle": bp.ID(),
+		}).Debug("SprayAndWait initialised new bundle from this host")
+	}
+}
+
+// SenderForBundle returns the Core's ConvergenceSenders.
+// If a node has more than 1 copy left it will send floor(copies/2) to the peer
+// and keep roof(copies/2) for itself
+func (bs BinarySpray) SenderForBundle(bp BundlePack) (css []cla.ConvergenceSender, del bool) {
+	metadata, _ := bs.bundleData[bp.ID()]
+	// if there are no copies left, we just wait until we meet the recipient
+	if metadata.remainingCopies < 2 {
+		log.WithFields(log.Fields{
+			"bundle": bp.ID(),
+		}).Debug("Not relaying bundle because there are copies left")
+		return nil, false
+	}
+
+	for _, cs := range bs.c.convergenceSenders {
+		var skip = false
+		for _, eid := range metadata.sent {
+			if cs.GetPeerEndpointID() == eid {
+				skip = true
+				break
+			}
+		}
+
+		if !skip {
+			css = append(css, cs)
+			metadata.sent = append(metadata.sent, cs.GetPeerEndpointID())
+
+			// we send half our remaining copies
+			sendCopies := metadata.remainingCopies / 2
+			metadata.remainingCopies = metadata.remainingCopies - sendCopies
+
+			// if the bundle already has a metadata-block
+			if metadataBlock, err := bp.Bundle.ExtensionBlock(ExtBlockTypeBinarySprayBlock); err == nil {
+				binarySprayBlock := metadataBlock.Value.(*BinarySprayBlock)
+				binarySprayBlock.SetCopies(sendCopies)
+			} else {
+				// if it doesn't, then create one
+				metadataBlock := NewBinarySprayBlock(sendCopies)
+				bp.Bundle.AddExtensionBlock(bundle.NewCanonicalBlock(0, 0, metadataBlock))
+			}
+
+			// we currently only send a bundle to a single peer at once
+			break
+		}
+	}
+
+	bs.bundleData[bp.ID()] = metadata
+
+	log.WithFields(log.Fields{
+		"bundle":              bp.ID(),
+		"convergence-senders": css,
+		"remaining copies":    metadata.remainingCopies,
+	}).Debug("BinarySpray selected Convergence Sender for an outgoing bundle")
+
+	del = false
+	return
+}
+
+// ReportFailure resets remaining copies if delivery was unsuccessful.
+func (bs BinarySpray) ReportFailure(bp BundlePack, sender cla.ConvergenceSender) {
+	log.WithFields(log.Fields{
+		"bundle":  bp.ID(),
+		"bad_cla": sender,
+	}).Debug("Transmission failure")
+
+	metadataBlock, _ := bp.Bundle.ExtensionBlock(ExtBlockTypeBinarySprayBlock)
+	binarySprayBlock := metadataBlock.Value.(*BinarySprayBlock)
+
+	metadata, _ := bs.bundleData[bp.ID()]
+	binarySprayBlock.SetCopies(metadata.remainingCopies + binarySprayBlock.RemainingCopies())
+
+	for i := 0; i < len(metadata.sent); i++ {
+		if metadata.sent[i] == sender.GetPeerEndpointID() {
+			metadata.sent = append(metadata.sent[:i], metadata.sent[i+1:]...)
+			break
+		}
+	}
+
+	bs.bundleData[bp.ID()] = metadata
+}
+
+const ExtBlockTypeBinarySprayBlock uint64 = 8
```

Comment:
The block type code 8 is already reserved for the Bundle Age Block, which is also implemented. Please select another block type code from the range between 192 and 255, as defined in section [4.2.3 of the BP7](https://tools.ietf.org/html/draft-ietf-dtn-bpbis-13#section-4.2.3).

Reply:
Done